scripts/import-open-api.ts

```diff
@@ -0,0 +1,236 @@
+          const schema = res.content["application/octet-stream"].schema!;
+          return isReference(schema) ? getRef(schema.$ref) : getScalar(schema);
+        } else {
+          return "void";
+        }
+      }
+    })
+    .join(` | `);
+
+/**
+ * Import and parse the openapi spec from a yaml/json
+ *
+ * @param path abosulte path of the yaml/json openapi 3.0.x file
+ */
+const importSpecs = (path: string): OpenAPIObject => {
+  const data = readFileSync(path, "utf-8");
+  const { ext } = parse(path);
+  return ext === ".yaml" ? YAML.parse(data) : JSON.parse(data);
+};
+
+/**
+ * Generate a restful-react Get compoment from openapi operation specs
+ *
+ * @param operation
+ */
+const generateGetComponent = (operation: OperationObject, verb: string, route: string) => {
+  if (!operation.operationId) {
+    throw new Error(`Every path must have a operationId - No operationId set for ${verb} ${route}`);
+  }
+
+  const componentName = pascal(operation.operationId!);
+
+  const isOk = ([statusCode]: [string, ResponseObject | ReferenceObject]) => statusCode.toString().startsWith("2");
+  const isError = ([statusCode]: [string, ResponseObject | ReferenceObject]) => !statusCode.toString().startsWith("2");
+
+  const responseTypes = getResponseTypes(Object.entries(operation.responses).filter(isOk), componentName);
+  const errorTypes = getResponseTypes(Object.entries(operation.responses).filter(isError), componentName);
+
+  return `
+export type ${componentName}Props = Omit<GetProps<${responseTypes}, ${errorTypes}>, "path">
+
+${operation.summary ? "// " + operation.summary : ""}
+export const ${componentName}: React.SFC<${componentName}Props> = (props) => (
+  <Get path="${route}" {...props} />
+)
+
+`;
+};
+
+/**
+ * Extract all types from #/components/schemas
+ *
+ * @param schemas
+ */
+const generateSchemaDefinition = (schemas: ComponentsObject["schemas"] = {}) =>
+  Object.entries(schemas)
+    .map(([name, schema]) => {
+      switch (schema.type || "object") {
+        case "object":
+          return `
+export interface ${pascal(name)} {
+${Object.entries(schema.properties || {})
+            .map(
+              ([key, properties]) =>
+                isReference(properties) ? `  ${key}: ${getRef(properties.$ref)}` : `${key}: ${getScalar(properties)}`,
+            )
+            .join("\n")}
+}
+`;
+        case "array":
+          return `export type ${pascal(name)} = ${getArray(schema)}`;
+        case "string":
+          return schema.enum
+            ? `export type ${pascal(name)} = "${schema.enum.join(`" | "`)}"`
+            : `export type ${pascal(name)} = string`;
+        default:
+          return `export type ${pascal(name)} = ${getScalar(schema)}`;
+      }
+    })
+    .join("\n\n");
+
+const main = (path: string) => {
+  const schema = importSpecs(path);
+
+  if (!schema.openapi.startsWith("3.0")) {
+    throw new Error("This tools can only parse open-api 3.0.x specifications");
```

Comment:
We _should_ support older versions too.

Reply:
Since a lot of tools can make the swagger -> open-api migration, it's not really a problem :wink: